Add new PyChop version and create model versioning

RastislavTuranyi opened the pull request:
Resolves the PyChop issue documented at [Mantid#38591](https://github.com/mantidproject/mantid/pull/38591)

Simultaneously, applies new versioning scheme for models, where all models are versioned with `_v{int}` at the end of their names to keep record of model changes, and aliases are created without the version to provide seamless access to the latest version of models.

RastislavTuranyi commented on src/resolution_functions/instrument.py:
```diff
@@ -643,8 +667,8 @@ def available_models_and_configurations(self) -> dict[str, list[str]]:
         models_and_configurations
             All models and all their configurations.
         """
-        return {model_name: list(model['configurations'].keys())
-                for model_name, model in self._models.items()}
+        return {model_name: list(self._resolve_model(model_name)['configurations'].keys())
+                for model_name in self._models.keys()}
```
With the current implementation, this is how this function would look like (the question whether it should return *all* models as it currently does, or only the non-versioned ones is separate), but I think the loop over `keys()` and then using the key to index into the dictionary is not the most elegant.

Perhaps we should create a new class (a `dict` subclass maybe) for storing the models dictionary that could handle the aliases nicer?

ajjackson replied:
I think the `.keys()` are redundant because `list` and `in` work with keys by default.

oerc0122 replied:
Given that Python handles things as pointers, dictionaries can just point to the same entity and be aliases.

RastislavTuranyi replied:
We have decided to turn this (and similar) property into ones that do not output a dictionary but a nicely formatted string - this way, the methods (such as `possible_configurations_for_model`) will be meant for API access by codes (when a type is required), and these properties will be meant for users trying to get information in REPL etc. This will reduce the need for special stuff with `self._models` (and in fact means we'll want to keep the distinction between specific models and aliases so that we can use it for formatting)

RastislavTuranyi replied:
This has now been implemented

oerc0122 commented on src/resolution_functions/instrument.py:
```diff
@@ -449,6 +446,33 @@ def _get_model_data(self, model_name: Optional[str] = None, **kwargs) -> tuple[M
+        except KeyError:
+            raise InvalidModelError(model_name, self)
+
+        if isinstance(model, str):
+            try:
+                model = self._models[model]
+            except KeyError:
+                raise InvalidModelError(model, self)
```
Could/should this recursively call `_resolve_model`? Would that be clearer?

RastislavTuranyi replied:
We have decided to enforce in the YAML data file spec that, if the model is a string, it *must* point to a model that has actual data, i.e. forbidding this kind of chaining.

ajjackson replied:
This eliminates some difficult (e.g. circular) cases at the acceptable cost of sometimes writing a few more characters in YAML.

ajjackson commented:
Can we xfail the appropriate tests so we get a green tick and verify nothing else was broken?

RastislavTuranyi commented:
I've changed the tests so that the old tests test the `PyChop_fit_v1` version specifically, and added new tests for testing `PyChop_fit_v2` which are mostly marked as XFAIL.

oerc0122 commented on src/resolution_functions/instrument.py:
```diff
@@ -685,12 +813,7 @@ def possible_configurations_for_model(self, model_name: str) -> list[str]:
-            raise InvalidModelError(model_name, self)
-
-        return list(model['configurations'].keys())
+        return list(self._resolve_model(model_name)['configurations'].keys())
```
`.keys()` is redundant

oerc0122 commented on src/resolution_functions/instrument.py:
```diff
@@ -809,16 +924,58 @@ def default_option_for_configuration(self, model_name: str, configuration: str)
+    table_str
+        A string containing the table.
+    """
+    longest = np.array(longest) + padding
+    out = []
+    separator = '|' + '|'.join(['-' * (i + 1) for i in longest]) + '|'
+
+    for i, line in enumerate(contents):
+        new_line = '| '
+        for val, length in zip(line, longest):
+            if val is None:
+                new_line += ' ' * length + '| '
+            else:
+                new_line += val + ' ' * (length - len(val)) + '| '
+
+        if line[0] is not None:
+            if i == 1:
+                out.append(separator.replace('-', '='))
+            else:
+                out.append(separator)
+
+        out.append(new_line)
+
+    out.append(separator)
+
+    return '\n'.join(out)
```
Maybe it's just my functional-programming brain, but you can avoid a lot of this fuss by using Python's format functions. It needs some hackery to get the *same* table, but a similar table could be done without incrementing `padding` and `preproc`, different formats, etc. `sep_nl` is just because you can't have `\` in an f-string. 
```suggestion
    padding = padding + 1
    preproc = [["-" if elem is None else elem for elem in row] for row in contents]

    sep_line = f"|{'|'.join('-' * (i+padding) for i in longest)}|"
    header_sep_line = sep_line.replace("-", "=")
    sep_nl = "\n" + sep_line

    data_line = "|" + "|".join(f"{{:^{length+padding}}}" for length in longest) + "|"
    formatted = map(lambda x: data_line.format(*x), preproc)

    out = f"""\
{sep_line}
{next(formatted)}
{header_sep_line}
{sep_nl.join(formatted)}
{sep_line}
    """

    return out
```

oerc0122 replied:
That's just because I stuck in centred (`^`) rather than left aligned (`<`) in the formats

oerc0122 replied:
Change `data_line` to:
```python
    data_line = "| " + "| ".join(f"{{:<{length+padding}}}" for length in longest) + "|"
```
And it should do what you had before.

oerc0122 replied:
Ah! I see what you mean, it was grouping things previously.

oerc0122 replied:
Makes `more_itertools` a required dep, but:
```suggestion
    padding = padding + 1

    preproc = [["" if elem is None else elem for elem in row] for row in contents]

    sep_line = f"|{'|'.join('-' * (i+padding) for i in longest)}|"
    header_sep_line = sep_line.replace("-", "=")
    sep_nl = "\n" + sep_line + "\n"
    data_line = "| " + "| ".join(f"{{:<{length+padding-1}}}" for length in longest) + "|"

    it = more_itertools.split_before(preproc, lambda x: x[0])
    formatted = ('\n'.join(map(lambda x: data_line.format(*x), blk)) for blk in it)

    out = f"""\
{sep_line}
{next(formatted)}
{header_sep_line}
{sep_nl.join(formatted)}
{sep_line}
    """

    return out
```

RastislavTuranyi replied:
That does indeed work! @ajjackson what do you think?

oerc0122 replied:
> I like the approach and it's easier to quickly see what is going on. Some more verbose variable names might help further. (e.g. `preproc -> contents_as_strings`) Maybe a comment above #966
> 
> ```
> # e.g. "|{:<10}|{:<10}|{:<12}|{:<16}|" for str.format()
> ```
> 
> Instead of adding the dependency we could implement a function here like:
> 
> ```python
> def split_on_model(contents: Iterator[list[str]]) -> Iterator[list[list[str]]]:
>     """Break the table lines into chunks where the first column is occupied"""
>     group = [next(contents)]
>     for line in contents:
>         if line[0]:
>             yield group
>             group = []
>         else:
>             group.append(line)
>     if group:
>         yield group
> ```
> 
> (I haven't tested this, typed straight into github website...)

You'll need `group = [line]` to avoid dropping the head.  That could also handle the preprocessing of blanks.

RastislavTuranyi replied:
So, do you want to go ahead with this? With both `split_on_model` and `_format_table` in a new utils module?

ajjackson replied:
Are they likely to be re-used? If so then collecting some generic output-printing utils makes sense. This kind of modularity is good for testing.

But if they would have private names and be used in one place, I'm not sure it's so beneficial to split them into a new module.

RastislavTuranyi replied:
Not at the moment, no. In the future, difficult to say, but I can't think of anything

RastislavTuranyi replied:
Could you give me an example of how the `split_on_model` is supposed to be used inside `_format_table`? Can't seem to figure out how it's supposed to replace `more_itertools.split_before`

oerc0122 commented on src/resolution_functions/instrument.py:
```diff
@@ -630,40 +674,124 @@ def available_models(self) -> list[str]:
+            A string containing the nicely formatted table.
+        """
+        contents = [['MODEL', 'ALIAS FOR', 'CONFIGURATIONS']]
+        longest_name, longest_alias, longest_config = 5, 9, 14
```
Makes it more obvious what these are.
```suggestion
        longest_name, longest_alias, longest_config = ​map(len, contents[0])
```

oerc0122 commented on src/resolution_functions/instrument.py:
```diff
@@ -630,40 +674,124 @@ def available_models(self) -> list[str]:
+        contents = [['MODEL', 'ALIAS FOR', 'CONFIGURATIONS']]
+        longest_name, longest_alias, longest_config = 5, 9, 14
+
+        for model_name, model_data in self._models.items():
+            length = len(model_name)
+            if length > longest_name:
+                longest_name = length
+
+            if isinstance(model_data, str):
+                contents.append([model_name, model_data, None])
+
+                length = len(model_data)
+                if length > longest_alias:
+                    longest_alias = length
+            else:
+                for i, config_name in enumerate(model_data['configurations']):
+                    if i == 0:
+                        contents.append([model_name, None, config_name])
+                    else:
+                        contents.append([None, None, config_name])
+
+                    length = len(config_name)
+                    if length > longest_config:
+                        longest_config = length
```
While it requires 2 runs through the elements, I would honestly not compute the length in this building-the-list stage as it makes the loop a lot more complex. 

I would iterate through the compiled list and compute the longest in each column after this. The list will be short (<1k elements), so the cost of multiple runs (unless this function is called a lot) is irrelevant. 

Pulling the logic out and using:
```
for name, alias, config in contents:
    longest_name  = max(longest_name, len(name))
    longest_alias  = max(longest_alias, len(alias))
    longest_config  = max(longest_config, len(config))
```
(or something more efficient) at the end makes the purpose clear.

RastislavTuranyi replied:
Maybe inside `_format_table`, we could do (though we would either need to do masking or change the `None` into e.g. `"-"`):

```
contents = np.array(contents)
longest = np.max(np.vectorize(len)(b), axis=0)
```

ajjackson replied:
Seems promising. It should be done after the None->str conversion, just in case we decide to replace None with `"----------"` 😉

RastislavTuranyi replied:
I think it'd be better to do that upstream, in `format_available_models_and_configurations` and `format_available_models_options`, and completely eliminite the `None`s as well as the pre-processing.

ajjackson replied:
With those sort of details already taken care of, it becomes more practical to have a generic table formatter that e.g. supports multiple criteria of how blocks are split. (Or maybe just requires the blocks to be pre-split as well.)

ajjackson commented on src/resolution_functions/instrument.py:
```diff
@@ -449,6 +448,33 @@ def _get_model_data(self, model_name: Optional[str] = None, **kwargs) -> tuple[M
+        try:
+            model = self._models[model_name]
+        except KeyError:
+            raise InvalidModelError(model_name, self)
```
Would it be appropriate for these to be chained exceptions? https://peps.python.org/pep-3134/

```python
except KeyError as e:
    raise InvalidModelError(model_name, self) from e
```
```

RastislavTuranyi replied:
I am not sure the additional information would be useful - the main purpose of this (and similar) exceptions is to communicate to the user that their input was wrong (and how to correct it), knowing that a `KeyError` was involved seems like unnecessary information. I don't think there would be any harm in the chaining, per se, so I am not necessarily opposed, but I also don't see the reason for - do you have any particular reason for this?

ajjackson replied:
I think it can give a cleaner output than "while handling X, Y was raised", but I'm not super familiar with it. Try raising both and see which output looks better?

RastislavTuranyi replied:
WITHOUT `from`:
```
Traceback (most recent call last):
  File "C:\Users\dni83241\Documents\Git\resolution_functions\src\resolution_functions\instrument.py", line 466, in _resolve_model
    model = self._models[model_name]
KeyError: 'PyChop_fit4'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "C:\Users\dni83241\AppData\Local\JetBrains\PyCharm Community Edition 2024.2.1\plugins\python-ce\helpers\pydev\pydevconsole.py", line 364, in runcode
    coro = func()
  File "<input>", line 1, in <module>
  File "C:\Users\dni83241\Documents\Git\resolution_functions\src\resolution_functions\instrument.py", line 468, in _resolve_model
    raise InvalidModelError(model_name, self)
resolution_functions.instrument.InvalidModelError: "PyChop_fit4" is not a valid model for the MAPS instrument version MAPS. This instrument only supports the following models: ['PyChop_fit'].
```

WITH `from`:
```
Traceback (most recent call last):
  File "C:\Users\dni83241\Documents\Git\resolution_functions\src\resolution_functions\instrument.py", line 466, in _resolve_model
    model = self._models[model_name]
KeyError: 'PyChop_fit4'
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "C:\Users\dni83241\AppData\Local\JetBrains\PyCharm Community Edition 2024.2.1\plugins\python-ce\helpers\pydev\pydevconsole.py", line 364, in runcode
    coro = func()
  File "<input>", line 1, in <module>
  File "C:\Users\dni83241\Documents\Git\resolution_functions\src\resolution_functions\instrument.py", line 468, in _resolve_model
    raise InvalidModelError(model_name, self) from e
resolution_functions.instrument.InvalidModelError: "PyChop_fit4" is not a valid model for the MAPS instrument version MAPS. This instrument only supports the following models: ['PyChop_fit'].
```

Now that you mention it, they are (almost) equally bad - I would prefer the `KeyError` not to show up at all (wasn't that the default behaviour?) - but the change in text in `from e` syntax makes it clearer that it is an intended behaviour, so we can go with that at least.

ajjackson replied:
It's more subtle than I expected! Chaining looks slightly more correct, but not a big deal really.

RastislavTuranyi replied:
Seems like the output I was hoping for can be obtained via `raise InvalidModelError() from None` , which gives:
```
Traceback (most recent call last):
  File "C:\Users\dni83241\AppData\Local\JetBrains\PyCharm Community Edition 2024.2.1\plugins\python-ce\helpers\pydev\pydevconsole.py", line 364, in runcode
    coro = func()
  File "<input>", line 1, in <module>
  File "C:\Users\dni83241\Documents\Git\resolution_functions\src\resolution_functions\instrument.py", line 468, in _resolve_model
    raise InvalidModelError(model_name, self) from None
resolution_functions.instrument.InvalidModelError: "invalid" is not a valid model for the ARCS instrument version ARCS. This instrument only supports the following models: ['PyChop_fit'].
```
any thoughts?

ajjackson replied:
This seems like the sort of thing `raise ... from None` is intended for, good call!

ajjackson commented on src/resolution_functions/instrument.py:
```diff
@@ -618,52 +644,154 @@ def get_model_signature(self, model_name: Optional[str] = None) -> Signature:
         A list of all :term:`models<model>` available for this :term:`version` of this
         :term:`instrument`.
 
+        Includes both all the versions of all models and
```
and ...?

RastislavTuranyi replied:
now updated

ajjackson commented on src/resolution_functions/instrument_data/arcs.yaml:
```diff
@@ -34,8 +33,8 @@ version:
-        configurations: &configurations
-            chopper_package:
+        configurations:
+            chopper_package: &configurations
```
What's the reasoning to move this rather than rename it as well?

RastislavTuranyi replied:
laziness, lack of attention to detail ...

ajjackson replied:
"the human condition"

RastislavTuranyi commented:
Everything previously suggested should now be implemented one way or another.

oerc0122 commented on src/resolution_functions/instrument.py:
```diff
@@ -355,7 +357,7 @@ def _get_file(instrument_name: str) -> tuple[str, Union[str, None]]:
             raise InvalidInstrumentError(
                 f'"{instrument_name}" is not a valid instrument name. Only the following instruments are '
-                f'supported: {list(INSTRUMENT_MAP.keys())}')
+                f'supported: {list(INSTRUMENT_MAP)}')
```
Could do
```suggestion
                f'supported: {', '.join(INSTRUMENT_MAP)}')
```
For slightly neater output. Fine as-is.

RastislavTuranyi replied:
I think I like having the brackets there a bit better, but maybe I'm weird

oerc0122 commented on src/resolution_functions/instrument.py:
```diff
@@ -618,52 +645,125 @@ def get_model_signature(self, model_name: Optional[str] = None) -> Signature:
         """
-        return {model_name: list(model['configurations'].keys())
-                for model_name, model in self._models.items()}
+        return [name for name, value in self._models.items() if not isinstance(value, str)]
```
This and the function above appear identical.

RastislavTuranyi replied:
Do you mean in the code or documentation?

The intention is to provide all possibilities for listing the available models:

- `available_unique_models` (this one) for listing the actual (versioned) models - the ones with data
- `available_models` for listing the aliased models (without versions) - the ones we recommend
- `all_available_models` for listing both of the above together

RastislavTuranyi replied:
Seems fine to me (in the same order):

- ` [name for name, value in self._models.items() if not isinstance(value, str)]`
- `[name for name, value in self._models.items() if isinstance(value, str)]`
- `list(self._models)`

gives output:

- `['PyChop_fit_v1', 'PyChop_fit_v2']`
- `['PyChop_fit']`
- `['PyChop_fit', 'PyChop_fit_v1', 'PyChop_fit_v2']`

as intended. Maybe it's the difference in `not isinstance` vs `isinstance`?

oerc0122 replied:
Ah! I missed the `not`, yes.

oerc0122 commented on src/resolution_functions/instrument.py:
```diff
@@ -618,52 +645,125 @@ def get_model_signature(self, model_name: Optional[str] = None) -> Signature:
+            if isinstance(model_data, str):
+                contents.append([model_name, model_data, ''])
+            else:
+                for i, config_name in enumerate(model_data['configurations']):
+                    if i == 0:
+                        contents.append([model_name, '', config_name])
+                    else:
+                        contents.append(['', '', config_name])
```
```suggestion
                tmp = [['', '', config_name]
                       for config_name in model_data['configurations']]
                tmp[0][0] = model_name
                contents.extend(tmp)
```
Is this more legible or less?

oerc0122 commented on tests/unit_tests/test_pychop.py:
```diff
@@ -81,6 +81,20 @@ def matrix_nonfermi_id(matrix_row: tuple[int, int, int]) -> str:
+    ['High-Resolution', 'High-Flux', 'SEQ-100-2.0-AST', 'SEQ-700-3.5-AST', 'ARCS-100-1.5-AST',
+     'ARCS-700-1.5-AST', 'ARCS-700-0.5-AST', 'ARCS-100-1.5-SMI', 'ARCS-700-1.5-SMI'],
+]
+INSTRUMENT_MATRIX_FERMI_V2 = list(
+        itertools.chain.from_iterable(
+            itertools.product(instr, configurations)
+                  for instr, configurations in zip(INSTRUMENTS_V2, INSTRUMENT_CONFIGURATIONS_V2)
+        )
+)
```
What is the logic here?
```
INSTRUMENT_MATRIX_FERMI_V2 = [
    pair for instr, configurations in zip(...)) for pair in product(instr, configurations)
]
```
?
Seems less magical.

RastislavTuranyi replied:
@ajjackson is better poised to answer that

ajjackson replied:
I'm not a big fan of double list comprehensions, I find them hard to read and really only use them for performance.

As Jacob has written it I _can_ tell that the second `for` is the "inner" loop because that's the only way it could work. It's not so bad; up to you @RastislavTuranyi

oerc0122 replied:
Nor am I, to be honest, it's one of the things I miss from other languages like e.g. Haskell
```haskell
cartProd :: [a] -> [b] -> [(a, b)]
cartProd xs ys = [(x, y) | x <- xs, y <- ys]

INSTRUMENT_MATRIX_FERMI_V2 = [pair | (a, b) <- zip INSTRUMENTS_V2 INSTRUMENT_CONFIGURATIONS_V2,
                                     pair <- cartProd a b]
```
Or even Raku with hyperoperators
```raku
$INSTRUMENT_MATRIX_FERMI_V2 = $INSTRUMENTS_V2 «X» $INSTRUMENT_CONFIGURATIONS_V2
```
Where:
- `X` is cart prod
-  `«f()»` applies the function drawing from each list, and `»` point to the shorter list (i.e. this says they should be of each length)

RastislavTuranyi replied:
I personally don't mind either way, so might as well leave it for now
